[BUG] Not importing correct paths

When deploying with runway 1.8.0, pre/post build/destroy hooks are not being imported correctly.

Steps to recreate using 
[runway-test1.zip](https://github.com/onicagroup/runway/files/4702987/runway-test1.zip):

```
pipenv install
make deploy
```
Results:
```
  File "<frozen importlib._bootstrap>", line 1006, in _gcd_import
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 953, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "<frozen importlib._bootstrap>", line 1006, in _gcd_import
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 962, in _find_and_load_unlocked
ModuleNotFoundError: No module named 'abc.hooks'; 'abc' is not a package
make: *** [deploy] Error 1
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -19,6 +19,7 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 
 ### Fixed
 - issue where `yamllint` and `cfnlint` could not be imported/executed from the Pyinstaller executables
+- fixed issue where CFNgin blueprints/hooks/lookups would encounter namespace collisions because imports were not being unloaded between instances
 
 ### Fixed
 - the friendly error when npm can't be found has returned

--- a/runway/cfngin/cfngin.py
+++ b/runway/cfngin/cfngin.py
@@ -6,7 +6,7 @@ import sys
 
 from yaml.constructor import ConstructorError
 
-from runway.util import MutableMap, argv, cached_property, environ
+from runway.util import MutableMap, SafeHaven, cached_property
 
 from .actions import build, destroy, diff
 from .config import render_parse_load as load_config
@@ -109,11 +109,11 @@ class CFNgin(object):
             sys_path = self.sys_path
         config_files = self.find_config_files(sys_path=sys_path)
 
-        with environ(self.__ctx.env_vars):
+        with SafeHaven(environ=self.__ctx.env_vars):
             for config in config_files:
                 ctx = self.load(config)
                 LOGGER.info('%s: deploying...', os.path.basename(config))
-                with argv('stacker', 'build', ctx.config_path):
+                with SafeHaven(argv=['stacker', 'build', ctx.config_path]):
                     action = build.Action(
                         context=ctx,
                         provider_builder=self._get_provider_builder(
@@ -141,11 +141,11 @@ class CFNgin(object):
         # destroy should run in reverse to handle dependencies
         config_files.reverse()
 
-        with environ(self.__ctx.env_vars):
+        with SafeHaven(environ=self.__ctx.env_vars):
             for config in config_files:
                 ctx = self.load(config)
                 LOGGER.info('%s: destroying...', os.path.basename(config))
-                with argv('stacker', 'destroy', ctx.config_path):
+                with SafeHaven(argv=['stacker', 'destroy', ctx.config_path]):
                     action = destroy.Action(
                         context=ctx,
                         provider_builder=self._get_provider_builder(
@@ -198,12 +198,12 @@ class CFNgin(object):
         if not sys_path:
             sys_path = self.sys_path
         config_files = self.find_config_files(sys_path=sys_path)
-        with environ(self.__ctx.env_vars):
+        with SafeHaven(environ=self.__ctx.env_vars):
             for config in config_files:
                 ctx = self.load(config)
                 LOGGER.info('%s: generating change sets...',
                             os.path.basename(config))
-                with argv('stacker', 'diff', ctx.config_path):
+                with SafeHaven(argv=['stacker', 'diff', ctx.config_path]):
                     action = diff.Action(
                         context=ctx,
                         provider_builder=self._get_provider_builder(

--- a/runway/util.py
+++ b/runway/util.py
@@ -1,17 +1,21 @@
 """Utility functions."""
 from __future__ import print_function
-from typing import Any, Dict, Iterator, List, Optional, Union  # noqa pylint: disable=unused-import
 
-from contextlib import contextmanager
 import hashlib
 import importlib
 import json
+import logging
 import os
 import platform
 import re
 import stat
-from subprocess import check_call
 import sys
+from contextlib import AbstractContextManager, contextmanager
+from copy import deepcopy
+from subprocess import check_call
+from typing import (Any, Dict, Iterator,  # noqa pylint: disable=unused-import
+                    List, Optional, Union)
+
 import six
 
 AWS_ENV_VARS = ('AWS_ACCESS_KEY_ID', 'AWS_SECRET_ACCESS_KEY',
@@ -279,6 +283,81 @@ class MutableMap(six.moves.collections_abc.MutableMapping):  # pylint: disable=n
         # type: () -> str
         """Return string representation of the object."""
         return json.dumps(self.data)
+
+
+class SafeHaven(AbstractContextManager):
+    """Context manager that caches and resets important values on exit.
+
+    Caches and resets os.environ, sys.argv, sys.modules, and sys.path.
+
+    """
+
+    # pylint: disable=redefined-outer-name
+    def __init__(self, argv=None, environ=None, sys_path=None):
+        """Instantiate class.
+
+        Args:
+            argv (Optional[List[str]]): Override the value of sys.argv.
+            environ (Optional[Dict[str, str]]): Override the value of os.environ.
+            sys_path (Optional[List[str]]): Override the value of sys.path.
+
+        """
+        self.__os_environ = deepcopy(os.environ)
+        self.__sys_argv = list(sys.argv)
+        # deepcopy can't pickle sys.modules and dict()/.copy() are not safe
+        self.__sys_modules = {k: v for k, v in sys.modules.items()}
+        self.__sys_path = list(sys.path)
+        self.log = logging.getLogger('runway.' + self.__class__.__name__)
+
+        if isinstance(argv, list):
+            sys.argv = argv
+        if isinstance(environ, dict):
+            os.environ = environ
+        if isinstance(sys_path, list):
+            sys.path = sys_path
+
+    def reset_all(self):
+        """Reset all values cached by this context manager."""
+        self.reset_os_environ()
+        self.reset_sys_argv()
+        self.reset_sys_modules()
+        self.reset_sys_path()
+
+    def reset_os_environ(self):
+        """Reset the value of os.environ."""
+        self.log.debug('resetting os.environ: %s', self.__os_environ)
+        os.environ = self.__os_environ
+
+    def reset_sys_argv(self):
+        """Reset the value of sys.argv."""
+        self.log.debug('resetting sys.argv: %s', self.__sys_argv)
+        sys.argv = self.__sys_argv
+
+    def reset_sys_modules(self):
+        """Reset the value of sys.modules."""
+        self.log.debug('resetting sys.modules: %s', self.__sys_modules)
+        for module in list(sys.modules.keys()):
+            if module not in self.__sys_modules:
+                self.log.debug('removed sys.module: {"%s": "%s"}', module,
+                               sys.modules.pop(module))
+
+    def reset_sys_path(self):
+        """Reset the value of sys.path."""
+        self.log.debug('resetting sys.path: %s', self.__sys_path)
+        sys.path = self.__sys_path
+
+    def __enter__(self):
+        """Enter the context manager.
+
+        Returns:
+            SafeHaven: Instance of the context manager.
+
+        """
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        """Exit the context manager."""
+        self.reset_all()
 
 
 @contextmanager

--- a/runway/util.py
+++ b/runway/util.py
@@ -11,7 +11,6 @@ import re
 import stat
 import sys
 from contextlib import AbstractContextManager, contextmanager
-from copy import deepcopy
 from subprocess import check_call
 from typing import (Any, Dict, Iterator,  # noqa pylint: disable=unused-import
                     List, Optional, Union)
@@ -298,11 +297,11 @@ class SafeHaven(AbstractContextManager):
 
         Args:
             argv (Optional[List[str]]): Override the value of sys.argv.
-            environ (Optional[Dict[str, str]]): Override the value of os.environ.
+            environ (Optional[Dict[str, str]]): Update os.environ.
             sys_path (Optional[List[str]]): Override the value of sys.path.
 
         """
-        self.__os_environ = deepcopy(os.environ)
+        self.__os_environ = dict(os.environ)
         self.__sys_argv = list(sys.argv)
         # deepcopy can't pickle sys.modules and dict()/.copy() are not safe
         # pylint: disable=unnecessary-comprehension
@@ -314,7 +313,7 @@ class SafeHaven(AbstractContextManager):
         if isinstance(argv, list):
             sys.argv = argv
         if isinstance(environ, dict):
-            os.environ = environ
+            os.environ.update(environ)
         if isinstance(sys_path, list):
             sys.path = sys_path
 
@@ -329,7 +328,8 @@ class SafeHaven(AbstractContextManager):
     def reset_os_environ(self):
         """Reset the value of os.environ."""
         self.log.debug('resetting os.environ: %s', self.__os_environ)
-        os.environ = self.__os_environ
+        os.environ.clear()
+        os.environ.update(self.__os_environ)
 
     def reset_sys_argv(self):
         """Reset the value of sys.argv."""
@@ -416,18 +416,16 @@ def environ(env=None, **kwargs):
     env = env or {}
     env.update(kwargs)
 
-    original_env = {key: os.getenv(key) for key in env}
+    original_env = dict(os.environ)
     os.environ.update(env)
 
     try:
+        print('Entered environ...')
         yield
     finally:
         # always restore original values
-        for key, val in original_env.items():
-            if val is None:
-                os.environ.pop(key, None)  # handle key missing
-            else:
-                os.environ[key] = val
+        os.environ.clear()
+        os.environ.update(original_env)
 
 
 def load_object_from_string(fqcn, try_reload=False):

--- a/runway/util.py
+++ b/runway/util.py
@@ -305,8 +305,10 @@ class SafeHaven(AbstractContextManager):
         self.__os_environ = deepcopy(os.environ)
         self.__sys_argv = list(sys.argv)
         # deepcopy can't pickle sys.modules and dict()/.copy() are not safe
+        # pylint: disable=unnecessary-comprehension
         self.__sys_modules = {k: v for k, v in sys.modules.items()}
         self.__sys_path = list(sys.path)
+        # more informative origin for log statements
         self.log = logging.getLogger('runway.' + self.__class__.__name__)
 
         if isinstance(argv, list):
@@ -318,6 +320,7 @@ class SafeHaven(AbstractContextManager):
 
     def reset_all(self):
         """Reset all values cached by this context manager."""
+        self.log.debug('resetting all managed values...')
         self.reset_os_environ()
         self.reset_sys_argv()
         self.reset_sys_modules()
@@ -336,6 +339,8 @@ class SafeHaven(AbstractContextManager):
     def reset_sys_modules(self):
         """Reset the value of sys.modules."""
         self.log.debug('resetting sys.modules: %s', self.__sys_modules)
+        # sys.modules can be manipulated to force reloading modules but,
+        # replacing it outright does not work as expected
         for module in list(sys.modules.keys()):
             if module not in self.__sys_modules:
                 self.log.debug('removed sys.module: {"%s": "%s"}', module,
@@ -353,10 +358,12 @@ class SafeHaven(AbstractContextManager):
             SafeHaven: Instance of the context manager.
 
         """
+        self.log.debug('entering a safe haven...')
         return self
 
     def __exit__(self, exc_type, exc_value, traceback):
         """Exit the context manager."""
+        self.log.debug('leaving the safe haven...')
         self.reset_all()
 
 

--- a/runway/util.py
+++ b/runway/util.py
@@ -425,7 +425,6 @@ def environ(env=None, **kwargs):
     os.environ.update(env)
 
     try:
-        print('Entered environ...')
         yield
     finally:
         # always restore original values

--- a/runway/util.py
+++ b/runway/util.py
@@ -10,12 +10,17 @@ import platform
 import re
 import stat
 import sys
-from contextlib import AbstractContextManager, contextmanager
+from contextlib import contextmanager
 from subprocess import check_call
 from typing import (Any, Dict, Iterator,  # noqa pylint: disable=unused-import
                     List, Optional, Union)
 
 import six
+
+if sys.version_info >= (3, 6):
+    from contextlib import AbstractContextManager  # pylint: disable=E
+else:
+    AbstractContextManager = object
 
 AWS_ENV_VARS = ('AWS_ACCESS_KEY_ID', 'AWS_SECRET_ACCESS_KEY',
                 'AWS_SESSION_TOKEN')

--- a/tests/cfngin/test_cfngin.py
+++ b/tests/cfngin/test_cfngin.py
@@ -1,10 +1,10 @@
 """Tests for runway.cfngin entry point."""
-# pylint: disable=no-self-use,protected-access
+# pylint: disable=no-self-use,protected-access.redefined-outer-name
 import os
 import shutil
 
 import pytest
-from mock import MagicMock, patch
+from mock import MagicMock, call, patch
 
 from runway.cfngin import CFNgin
 from runway.context import Context
@@ -21,6 +21,15 @@ def copy_basic_fixtures(cfngin_fixtures, tmp_path):
                  dest=tmp_path / 'test-us-east-1.env')
     copy_fixture(src=cfngin_fixtures / 'configs' / 'basic.yml',
                  dest=tmp_path / 'basic.yml')
+
+
+@pytest.fixture(scope='function')
+def patch_safehaven(monkeypatch):
+    """Patch SafeHaven."""
+    mock_haven = MagicMock()
+    mock_haven.return_value = mock_haven
+    monkeypatch.setattr('runway.cfngin.cfngin.SafeHaven', mock_haven)
+    return mock_haven
 
 
 class TestCFNgin(object):
@@ -77,7 +86,8 @@ class TestCFNgin(object):
         assert result.env_file.test_value == 'lab-ca-central-1'
 
     @patch('runway.cfngin.actions.build.Action')
-    def test_deploy(self, mock_action, cfngin_fixtures, tmp_path):
+    def test_deploy(self, mock_action, cfngin_fixtures, tmp_path,
+                    patch_safehaven):
         """Test deploy with two files & class init."""
         mock_instance = self.configure_mock_action_instance(mock_action)
         copy_basic_fixtures(cfngin_fixtures, tmp_path)
@@ -110,21 +120,41 @@ class TestCFNgin(object):
                                           'tail': False},
                                          {'concurrency': 0,
                                           'tail': False}])
+        patch_safehaven.assert_has_calls([
+            call(environ=context.env_vars),
+            call.__enter__(),
+            call(argv=['stacker', 'build', str(tmp_path / 'basic.yml')]),
+            call.__enter__(),
+            call.__exit__(None, None, None),
+            call(argv=['stacker', 'build', str(tmp_path / 'basic2.yml')]),
+            call.__enter__(),
+            call.__exit__(None, None, None),
+            call.__exit__(None, None, None),
+        ])
 
     @patch('runway.cfngin.actions.destroy.Action')
-    def test_destroy(self, mock_action, cfngin_fixtures, tmp_path):
+    def test_destroy(self, mock_action, cfngin_fixtures, tmp_path,
+                     patch_safehaven):
         """Test destroy."""
         mock_instance = self.configure_mock_action_instance(mock_action)
         copy_basic_fixtures(cfngin_fixtures, tmp_path)
 
-        # support python < 3.6
-        cfngin = CFNgin(ctx=self.get_context(), sys_path=str(tmp_path))
+        context = self.get_context()
+        cfngin = CFNgin(ctx=context, sys_path=str(tmp_path))
         cfngin.destroy()
 
         mock_action.assert_called_once()
         mock_instance.execute.assert_called_once_with(concurrency=0,
                                                       force=True,
                                                       tail=False)
+        patch_safehaven.assert_has_calls([
+            call(environ=context.env_vars),
+            call.__enter__(),
+            call(argv=['stacker', 'destroy', str(tmp_path / 'basic.yml')]),
+            call.__enter__(),
+            call.__exit__(None, None, None),
+            call.__exit__(None, None, None),
+        ])
 
     def test_load(self, cfngin_fixtures, tmp_path):
         """Test load."""
@@ -152,17 +182,26 @@ class TestCFNgin(object):
         assert 'appears to be a CloudFormation template' in caplog.text
 
     @patch('runway.cfngin.actions.diff.Action')
-    def test_plan(self, mock_action, cfngin_fixtures, tmp_path):
+    def test_plan(self, mock_action, cfngin_fixtures, tmp_path,
+                  patch_safehaven):
         """Test plan."""
         mock_instance = self.configure_mock_action_instance(mock_action)
         copy_basic_fixtures(cfngin_fixtures, tmp_path)
 
-        # support python < 3.6
-        cfngin = CFNgin(ctx=self.get_context(), sys_path=str(tmp_path))
+        context = self.get_context()
+        cfngin = CFNgin(ctx=context, sys_path=str(tmp_path))
         cfngin.plan()
 
         mock_action.assert_called_once()
         mock_instance.execute.assert_called_once_with()
+        patch_safehaven.assert_has_calls([
+            call(environ=context.env_vars),
+            call.__enter__(),
+            call(argv=['stacker', 'diff', str(tmp_path / 'basic.yml')]),
+            call.__enter__(),
+            call.__exit__(None, None, None),
+            call.__exit__(None, None, None),
+        ])
 
     def test_should_skip(self, cfngin_fixtures, tmp_path):
         """Test should_skip."""

--- a/tests/fixtures/empty_module.py
+++ b/tests/fixtures/empty_module.py
@@ -1,0 +1,5 @@
+"""Used to test SafeHaven sys.modules reset.
+
+!!DO NOT IMPORT FOR ANY OTHER TEST UNLESS IT UNLOADS THIS MODULE ON TEARDOWN!!
+
+"""

--- a/tests/test_util.py
+++ b/tests/test_util.py
@@ -1,13 +1,16 @@
 """Test Runway utils."""
 # pylint: disable=no-self-use
+import logging
 import os
 import string
 import sys
 
+import pytest
 from mock import MagicMock, patch
+from runway.util import (MutableMap, SafeHaven, argv, environ,
+                         load_object_from_string)
 
-from runway.util import MutableMap, argv, environ, load_object_from_string
-
+MODULE = 'runway.util'
 VALUE = {
     'bool_val': False,
     'dict_val': {'test': 'success'},
@@ -85,6 +88,127 @@ class TestMutableMap:
             'default_val', 'default should be used'
         assert mute_map.find('str_val', 'default_val') == \
             VALUE['str_val'], 'default should be ignored'
+
+
+class TestSafeHaven(object):
+    """Test SafeHaven context manager."""
+
+    TEST_PARAMS = [
+        (None), ('string'), ({}), ({'TEST_KEY': 'TEST_VAL'}),
+        (['runway', 'test'])
+    ]
+
+    def test_context_manager_magic(self, caplog, monkeypatch):
+        """Test init and the attributes it sets."""
+        mock_reset_all = MagicMock()
+        caplog.set_level(logging.DEBUG, 'runway.SafeHaven')
+        monkeypatch.setattr(MODULE + '.os', MagicMock())
+        monkeypatch.setattr(MODULE + '.sys', MagicMock())
+        monkeypatch.setattr(SafeHaven, 'reset_all', mock_reset_all)
+
+        with SafeHaven() as result:
+            assert isinstance(result, SafeHaven)
+        mock_reset_all.assert_called_once()
+        assert caplog.messages == ['entering a safe haven...',
+                                   'leaving the safe haven...']
+
+    @pytest.mark.parametrize('provided', TEST_PARAMS)
+    def test_os_environ(self, provided, caplog, monkeypatch):
+        """Test os.environ interactions."""
+        caplog.set_level(logging.DEBUG, 'runway.SafeHaven')
+        monkeypatch.setattr(SafeHaven, 'reset_all', MagicMock())
+
+        orig_val = dict(os.environ)
+        expected_val = dict(os.environ)
+        expected_logs = ['entering a safe haven...',
+                         'resetting os.environ: %s' % orig_val,
+                         'leaving the safe haven...']
+
+        if isinstance(provided, dict):
+            expected_val.update(provided)
+
+        with SafeHaven(environ=provided) as obj:
+            assert os.environ == expected_val
+            os.environ.update({'SOMETHING_ELSE': 'val'})
+            obj.reset_os_environ()
+        assert os.environ == orig_val
+        assert caplog.messages == expected_logs
+
+    def test_reset_all(self, caplog, monkeypatch):
+        """Test reset_all."""
+        mock_method = MagicMock()
+        caplog.set_level(logging.DEBUG, 'runway.SafeHaven')
+        monkeypatch.setattr(SafeHaven, 'reset_os_environ', mock_method)
+        monkeypatch.setattr(SafeHaven, 'reset_sys_argv', mock_method)
+        monkeypatch.setattr(SafeHaven, 'reset_sys_modules', mock_method)
+        monkeypatch.setattr(SafeHaven, 'reset_sys_path', mock_method)
+
+        expected_logs = ['entering a safe haven...',
+                         'resetting all managed values...',
+                         'leaving the safe haven...',
+                         'resetting all managed values...']
+
+        with SafeHaven() as obj:
+            obj.reset_all()
+            assert mock_method.call_count == 4
+        assert mock_method.call_count == 8  # called again on exit
+        assert caplog.messages == expected_logs
+
+    @pytest.mark.parametrize('provided', TEST_PARAMS)
+    def test_sys_argv(self, provided, caplog, monkeypatch):
+        """Test sys.argv interactions."""
+        caplog.set_level(logging.DEBUG, 'runway.SafeHaven')
+        monkeypatch.setattr(SafeHaven, 'reset_all', MagicMock())
+
+        orig_val = list(sys.argv)
+        expected_val = provided if isinstance(provided, list) else list(sys.argv)
+        expected_logs = ['entering a safe haven...',
+                         'resetting sys.argv: %s' % orig_val,
+                         'leaving the safe haven...']
+
+        with SafeHaven(argv=provided) as obj:
+            assert sys.argv == expected_val
+            sys.argv.append('something-else')
+            obj.reset_sys_argv()
+        assert sys.argv == orig_val
+        assert caplog.messages == expected_logs
+
+    def test_sys_modules(self, caplog, monkeypatch):
+        """Test sys.modules interactions."""
+        caplog.set_level(logging.DEBUG, 'runway.SafeHaven')
+        monkeypatch.setattr(SafeHaven, 'reset_all', MagicMock())
+
+        # pylint: disable=unnecessary-comprehension
+        orig_val = {k: v for k, v in sys.modules.items()}
+        expected_logs = ['entering a safe haven...',
+                         'resetting sys.modules: %s' % orig_val]
+
+        with SafeHaven() as obj:
+            from .fixtures import mock_hooks  # noqa pylint: disable=E,W,C
+            assert sys.modules != orig_val
+            obj.reset_sys_modules()
+        assert sys.modules == orig_val
+        assert caplog.messages[:2] == expected_logs
+        assert caplog.messages[-1] == 'leaving the safe haven...'
+
+    @pytest.mark.parametrize('provided', TEST_PARAMS)
+    def test_sys_path(self, provided, caplog, monkeypatch):
+        """Test sys.path interactions."""
+        caplog.set_level(logging.DEBUG, 'runway.SafeHaven')
+        monkeypatch.setattr(SafeHaven, 'reset_all', MagicMock())
+
+        orig_val = list(sys.path)
+        expected_val = provided if isinstance(provided, list) else list(sys.path)
+        expected_logs = ['entering a safe haven...',
+                         'resetting sys.path: %s' % orig_val,
+                         'leaving the safe haven...']
+
+        with SafeHaven(sys_path=provided) as obj:
+            assert sys.path == expected_val
+            sys.path.append('something-else')
+            obj.reset_sys_path()
+        assert sys.path == orig_val
+        assert caplog.messages == expected_logs
 
 
 @patch.object(sys, 'argv', ['runway', 'deploy'])

--- a/tests/test_util.py
+++ b/tests/test_util.py
@@ -101,16 +101,17 @@ def test_argv():
     assert sys.argv == orig_expected, 'validate value returned to original'
 
 
-@patch.object(os, 'environ', {'TEST_PARAM': 'initial value'})
 def test_environ():
     """Test environ."""
-    orig_expected = {'TEST_PARAM': 'initial value'}
+    orig_expected = dict(os.environ)
     override = {'TEST_PARAM': 'override', 'new_param': 'value'}
+    override_expected = dict(orig_expected)
+    override_expected.update(override)
 
     assert os.environ == orig_expected, 'validate original value'
 
     with environ(override):
-        assert os.environ == override, 'validate override'
+        assert os.environ == override_expected, 'validate override'
         assert os.environ.pop('new_param') == 'value'
 
     assert os.environ == orig_expected, 'validate value returned to original'

